### DISCUSSION TITLE
Debug data for promises

### DISCUSSION THREAD

**Discussion by EligiusSantori** [reactphp/reactphp#551] [Q&A] [answered]
Please add something to help debugging the code written via promises. Closues-based code makes stacktrace useless, and without stacktrace debugging some mostly working remote code is barely possible. On Node.js there is whole layer of knowledge exists to cope with it. I think I also saw there custom promises implementation which solve it in some way.

**Comment by EligiusSantori** [comment]
I guess simplest way to do it is using PHP 8 attributes. Then code may look like:

```php
$client = new React\Http\Browser();

$client->get('http://www.google.com/')->then(#[StepName('handle-response')] function (Psr\Http\Message\ResponseInterface $response) {
    throw new \Exception('Dummy exception.');
})->catch(function (\Throwable $reason, ?string $stepName) {
    file_put_contents(STDERR, "Exception: ".$reason->getMessage()." in step ".$stepName);
});
```

Or in more general way:
```php
$client = new React\Http\Browser();

$client->get('http://www.google.com/')->then(#[StepName('handle-response')] function (Psr\Http\Message\ResponseInterface $response) {
    throw new \Exception('Dummy exception.');
})->catch(function (\Throwable $reason, ?Closure $fn) {
    //...
});
```

**Comment by SimonFrings** [comment, marked as answer]
Hey @EligiusSantori, I can understand that this makes working with promises, especially when it comes to debugging, somewhat more complex. This is one of the reasons why we released v4 of our [reactphp/asnyc package](https://github.com/reactphp/async), so you can simply use `async()` and `await()` on PHP 8.1+ without caring too much about promises. If you're using a PHP version below 8.1, you can also use coroutines, which work in a similar way. Additionally, the more we approach the future of ReactPHP (ReactPHP v3 & v4), the less you have to deal with promises in general.

With this all said, I think it would make more sense to invest the time to move towards a ReactPHP v3/v4, as promises will move more and more into the background.

Thanks for bringing in your idea, hope this gives some insights into our plans for ReactPHP and helps you proceed  :)

**Reply by EligiusSantori** [reply]
Hello, @SimonFrings. Thanks for your response, I'll try async/await approach in the next few days and will report here is it solves the problem. As I read before, JavaScript has async/await support in it's core and it also has some improvements in exceptions handling related to async/await, so I didn't expect much from library-level implementation. I'll follow your recommendation, but I need some time to try it and give a thoughtful answer.

**Reply by EligiusSantori** [reply]
UPD: I tried async/await and it works good enough both for user-code exceptions and deep nested exceptions in promise-using libraries. The problem can be considered solved. Thanks again.
